routes/api/project/location/vm.rb

```diff
@@ -5,49 +5,115 @@ class CloverApi
+        allowed_optional_parameters = ["size", "unix_user", "boot_image", "enable_ip4", "private_subnet_id"]
+
+        request_body_params = Validation.validate_request_body(r.body.read, required_parameters, allowed_optional_parameters)
 
-      unless vm
-        response.status = 404
-        r.halt
+        if request_body_params["private_subnet_id"]
+          ps_id = UBID.parse(request_body_params["private_subnet_id"]).to_uuid
+          Authorization.authorize(@current_user.id, "PrivateSubnet:view", ps_id)
+          request_body_params.delete("private_subnet_id")
+        end
+
+        st = Prog::Vm::Nexus.assemble(
+          request_body_params["public_key"],
+          @project.id,
+          name: vm_name,
+          location: @location,
+          private_subnet_id: ps_id,
+          **request_body_params.except(*required_parameters).transform_keys { |key| key.to_sym }
+        )
+
+        serialize(st.subject)
       end
 
-      r.get true do
-        Authorization.authorize(@current_user.id, "Vm:view", vm.id)
+      vm = @project.vms_dataset.where(location: @location).where { {Sequel[:vm][:name] => vm_name} }.first
+      handle_vm_requests(@current_user, vm)
+    end
+  end
+
+  def handle_vm_requests(user, vm)
+    unless vm
+      response.status = request.delete? ? 204 : 404
+      request.halt
+    end
+
+    request.get true do
+      Authorization.authorize(user.id, "Vm:view", vm.id)
+      serialize(vm)
+    end
+
+    request.delete true do
+      Authorization.authorize(user.id, "Vm:delete", vm.id)
+      vm.incr_destroy
+
+      response.status = 204
+      request.halt
+    end
+
+    request.on "firewall-rule" do
+      request.post true do
+        Authorization.authorize(user.id, "Vm:Firewall:edit", vm.id)
+
+        required_parameters = ["cidr"]
+        allowed_optional_parameters = ["port_range"]
+
+        request_body_params = Validation.validate_request_body(request.body.read, required_parameters, allowed_optional_parameters)
+
+        Validation.validate_cidr(request_body_params["cidr"])
+        port_range = if request_body_params["port_range"].nil?
+          [0, 65535]
+        else
+          Validation.validate_port_range(request_body_params["port_range"])
+          request_body_params["port_range"].split("..").map(&:to_i)
+        end
+
+        pg_range = Sequel.pg_range(port_range.first..port_range.last)
+
+        vm.firewalls.first.insert_firewall_rule(request_body_params["cidr"], pg_range)
 
         serialize(vm)
       end
 
-      r.delete true do
-        Authorization.authorize(@current_user.id, "Vm:delete", vm.id)
+      request.is String do |firewall_rule_ubid|
+        request.delete true do
+          Authorization.authorize(user.id, "Vm:Firewall:edit", vm.id)
+          fwr = FirewallRule.from_ubid(firewall_rule_ubid)
 
-        vm.incr_destroy
+          if fwr
+            fwr.destroy
+            vm.incr_update_firewall_rules
+          end
```

Comment:
nit: _(it includes above line, GitHub doesn't allow to select it)_
```suggestion
          if (fwr = FirewallRule.from_ubid(firewall_rule_ubid))
            fwr.destroy
            vm.incr_update_firewall_rules
          end
```